pallets/connectors-gateway/src/lib.rs

```diff
@@ -0,0 +1,277 @@
+	#[pallet::generate_deposit(pub (super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// The router for a given domain was set.
+		DomainRouterSet { domain: Domain, router: T::Router },
+
+		/// A connector was added to a domain.
+		ConnectorAdded { connector: DomainAddress },
+
+		/// A connector was removed from a domain.
+		ConnectorRemoved { connector: DomainAddress },
+	}
+
+	/// Storage for domain routers.
+	///
+	/// This can only be set by an admin.
+	#[pallet::storage]
+	pub(crate) type DomainRouters<T: Config> = StorageMap<_, Blake2_128Concat, Domain, T::Router>;
+
+	/// Storage that contains a limited number of whitelisted connectors for a
+	/// particular domain.
+	///
+	/// This can only be modified by an admin.
+	#[pallet::storage]
+	pub(crate) type ConnectorsAllowlist<T: Config> = StorageMap<
+		_,
+		Blake2_128Concat,
+		Domain,
+		BoundedVec<DomainAddress, T::MaxConnectorsPerDomain>,
+		ValueQuery,
+	>;
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Router initialization failed.
+		RouterInitFailed,
+
+		/// The origin of the message to be processed is invalid.
+		InvalidMessageOrigin,
+
+		/// The domain is not supported.
+		DomainNotSupported,
+
+		/// Message decoding error.
+		MessageDecode,
+
+		/// Connector was already added to the domain.
+		ConnectorAlreadyAdded,
+
+		/// Maximum number of connectors for a domain was reached.
+		MaxConnectorsReached,
+
+		/// Connector was not found.
+		ConnectorNotFound,
+
+		/// Unknown connector.
+		UnknownConnector,
+
+		/// Router not found.
+		RouterNotFound,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Set a domain's router,
+		#[pallet::weight(< T as Config >::WeightInfo::set_domain_router())]
+		#[pallet::call_index(0)]
+		pub fn set_domain_router(
+			origin: OriginFor<T>,
+			domain: Domain,
+			router: T::Router,
+		) -> DispatchResult {
+			T::AdminOrigin::ensure_origin(origin)?;
+
+			ensure!(domain != Domain::Centrifuge, Error::<T>::DomainNotSupported);
+
+			router.init().map_err(|_| Error::<T>::RouterInitFailed)?;
+
+			<DomainRouters<T>>::insert(domain.clone(), router.clone());
+
+			Self::deposit_event(Event::DomainRouterSet { domain, router });
+
+			Ok(())
+		}
+
+		/// Add a connector for a specific domain.
+		#[pallet::weight(< T as Config >::WeightInfo::add_connector())]
+		#[pallet::call_index(1)]
+		pub fn add_connector(origin: OriginFor<T>, connector: DomainAddress) -> DispatchResult {
+			T::AdminOrigin::ensure_origin(origin)?;
+
+			ensure!(
+				connector.domain() != Domain::Centrifuge,
+				Error::<T>::DomainNotSupported
+			);
+
+			<ConnectorsAllowlist<T>>::try_mutate(connector.domain(), |connectors| {
+				if connectors.iter().find(|s| s.eq(&&connector)).is_some() {
+					return Err(Error::<T>::ConnectorAlreadyAdded.into());
+				}
+
+				connectors
+					.try_push(connector.clone())
+					.map_err(|_| Error::<T>::MaxConnectorsReached)?;
+
+				Self::deposit_event(Event::ConnectorAdded { connector });
+
+				Ok(())
+			})
```

Comment:
```suggestion
				ensure!(
				  !<ConnectorsAllowlist<T>>:contains(connector.domain(), connector), 
				  Error::<T>::ConnectorAlreadyAdded
				);
     
				<ConnectorsAllowlist<T>>::insert(connector.domain(), connector, ());
				Self::deposit_event(Event::ConnectorAdded { connector });

				Ok(())
			})
```

Reply:
It would be even better if `insert` fails if it already exists, but as I looked into the code I couldn't find that error case being handled there

Reply:
Done.